presto-main/src/main/java/com/facebook/presto/sql/analyzer/FunctionArgumentCheckerForAccessControlUtils.java

```diff
@@ -0,0 +1,170 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.toIntExact;
+import static java.util.Collections.reverse;
+
+public class FunctionArgumentCheckerForAccessControlUtils
+{
+    private static final QualifiedName TRANSFORM = QualifiedName.of("transform");
+
+    private FunctionArgumentCheckerForAccessControlUtils() {}
+
+    // Returns whether function argument at `argumentIndex` for function `node` needs to be checked
+    // for column level access control.
+    // For e.g., consider SQL `transform(arr, col -> col.x)`
+    // Here, we only need to check for access of subfield `x` in column `arr` which is of type `Array<struct>`.
+    // So we can just parse lambda and ignore the first argument for access checks.
+    public static boolean isUnusedArgumentForAccessControl(FunctionCall node, int argumentIndex, ExpressionAnalyzer.Context context)
+    {
+        if (node.getName().equals(TRANSFORM)) {
+            checkState(node.getArguments().size() == 2);
+            return argumentIndex == 0;
+        }
+        return false;
+    }
+
+    // Parses arguments of function `node` which are a lambda expression, and returns a map
+    // of their lambda arguments to resolved subfield.
+    // For e.g., consider SQL `SELECT transform(arr, col -> col.x) FROM table`
+    // Return value = Map('col' -> ResolvedSubfield(table.arr))
+    public static Map<Identifier, ResolvedSubfield> getResolvedLambdaArguments(
+            FunctionCall node,
+            StackableAstVisitorContext<ExpressionAnalyzer.Context> context,
+            Map<NodeRef<Expression>, Type> expressionTypes)
+    {
+        ImmutableMap.Builder<Identifier, ResolvedSubfield> resolvedLambdaArguments = ImmutableMap.builder();
+        if (node.getName().equals(TRANSFORM)) {
+            checkState(node.getArguments().size() == 2);
+            if (!(node.getArguments().get(1) instanceof LambdaExpression)) {
+                return ImmutableMap.of();
+            }
+            Expression arrayExpression = node.getArguments().get(0);
+            LambdaExpression lambdaExpression = ((LambdaExpression) node.getArguments().get(1));
+            Optional<ResolvedSubfield> resolvedSubfield = resolveSubfield(arrayExpression, context, expressionTypes);
+            if (resolvedSubfield.isPresent()) {
+                resolvedLambdaArguments.put(
+                        lambdaExpression.getArguments().get(0).getName(),
+                        resolvedSubfield.get());
+            }
+        }
+        return resolvedLambdaArguments.build();
+    }
+
+    public static Optional<ResolvedSubfield> resolveSubfield(
+            Expression node,
+            StackableAstVisitorContext<ExpressionAnalyzer.Context> context,
+            Map<NodeRef<Expression>, Type> expressionTypes)
+    {
+        // If expression is nested with multiple dereferences and subscripts, we only look at the topmost one.
```

Comment:
why?  why not use recursion to go into all of them?

Reply:
This part is moved from ExpressionAnalyzer as well. This function is called once for each dereference, when multiple dereferences are nested. So, we have this check that we only look at the dereference chain once.